Initial spin-button a11y commit (#24075)
microsoft/fluentui: packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md

| Down arrow | Increments value lower, based on the step prop (defaulting to 1) |
| Page up | Increments value higher, based on the stepPage prop (defaulting to 1) |
| Page down | Increments value lower, based on the stepPage prop (defaulting to 1) |
| Typing valid value | Valid value |

### High contrast mode

Be careful about visibility of the spin button arrows in high contrast mode.
TODO: Either add specifics about SVGs in high contrast and how to customize the icon color without messing up HCM or (ideally) link to the Icon section on high contrast, when it exists.

### Motion and animation

There is an animation when this element receives focus. This respects reduced-motion media queries.

When a user holds down an edit key or mouse click, there is an animation of spinning numbers to provide a sense of how fast the value is changing. This animation is an interaction cue and therefore is acceptable even when a user has set a preference for reduced motion.

### Semantics

## Known issues

Placeholder: If there are any current known bugs with browsers or AT that impact this component, list them here.

Mouse press quick changes to value are not announced to screen readers.

> 

Reviewer: Some known issues.

### Narrator + Edge

1. When no `min` or `max` values are set Narrator announces "minimum 0" and "maximum 0". This is misleading because when these values are not set `SpinButton` does not enforce a min or max value. [Example](https://codesandbox.io/s/sample-spinbutton-zxy72x).

### NVDA + Edge/Chrome

1. When focused on the `SpinButton` input field pressing up/down arrows announces "blank". NVDA + Firefox announces the correct value.

## Usage

### When to choose Spin button

SpinButtons allow someone to incrementally adjust a value in small steps.

SpinButtons are a better choice than Input and Input with type: number when clear indication is needed that there are maximum and minimum allowed values.

SpinButtons are a better choice than Slider when there are many valid values and Slider would not provide enough granularity to choose a precise value.

### Required props

> Placeholder: If there are any props or other specific work that the author must do to make the default usage of the component accessible, list those here. One example is that all form components must have a label provided by the author. If the component has any internally defined strings that should be localized, list those here.

Reviewer: Authors must provide a label for `SpinButton`. The recommended pattern for Fluent inputs is to use the [`Label`](https://react.fluentui.dev/?path=/docs/components-label--default) component like this:

```jsx
<Label htmlFor="example-spinbutton">A SpinButton</Label>
<SpinButton id="example-spinbutton" defaultValue={10} />
```

[Here's a live example](https://codesandbox.io/s/sample-spinbutton-zxy72x?file=/src/index.tsx:532-639).

> Open question: How does everything read? Name, role, value? What about when there are prefixes or postfixes?

Reviewer: How the SpinButton is read depends on the screen reader used but generally it reads (see comment above about known issues):

1. Label
2. "spin button"
3. "value" where "value" is the current displayed value

> What about when there are prefixes or postfixes?

Prefixes, postfixes and other display formatted (e.g., mapping the numbers 1-12 to January-December) is handled by setting the `aria-valuetext` attribute.

How this is read also depends on the screen reader used but a screen reader should read the displayed text consistently with how it would read it in other contexts (i.e., "$1" should be read as "one dollar"). [Example](https://react.fluentui.dev/?path=/docs/components-spinbutton--default#display-value).

### Styling

### Advanced usage

Placeholder: Cover potential use cases not included in our storybook examples here (or even included storybook examples, if they're complex require a notable amout of work or nuanced understanding from authors)

#### Child content restrictions

The component has the following structure:
* Input field
* Up button

>   * Up carat symbol

Reviewer: What do you think about?
```suggestion
  * Increment icon
```

* Down button

>   * Down carat symbol

Reviewer: and
```suggestion
  * Decrement icon
```

> #### Component-specific usage warnings

Reviewer: These seem like they belong in the "Known Issues" section.

## Component features and behavior

This section documents accessibility-related behavoirs of spin buttons.

### Keyboarding

#### Tab Order
1. Value field

#### States
1. Committed (Focused and value committed)
2. Editing (Focused and editing)

#### Keyboard State Diagram

| Starting state | Transition | Resulting state |
| ---------- | ------------ | ------- |
| Content before spin button | Tab | Committed |
| Committed  | Tab | Content after spin button (Value committed) |
| Editing  | Tab | Content after spin button (Value committed) |
| Editing (Autofill option provided)  | Tab | Content after spin button (Value committed) |
| Editing  | Enter | Committed |
| Committed  | Any Edit Key (that results in a change) | Editing |
| Editing  | Any Edit Key | Editing |
| Content after spin button (Value committed) | Shift + Tab | Committed |

##### Edit keys
| Edit Key | Result |
| ---------- | ------------ |
| Home | First item in defined range |
| End | Last item in defined range |

> | Up arrow | One increment higher |

Reviewer: The `step` prop controls the increment/decrement amount. It defaults to `1` but it could be `10`. What do you think about changing the results cell to

```
`step` increments higher
```

or

```
Increments by small about. Defaults to `1` but can be changed with `step` prop.
```

| Down arrow | One increment lower |

> | Page up | Several increments higher |

Reviewer: Like the up/down arrows this value is controlled by `stepPage`. It also defaults to `1` but can be set to any numeric value.

| Page down | Several increments lower |
| Typing valid value | Valid value |

> | Begin typing valid value | Autofill option provided |

Author: It's in the [Figma Spec](https://www.figma.com/proto/4ErMjZ0Mi72legBtMAbbEp/SpinButton?node-id=309%3A57799). We could try to discourage non-numeric numbers by removing that section.

> | Begin typing valid value | Autofill option provided |

Reviewer: We intentionally omitted auto complete functionality for now because, as Sarah noted, it's not clearly a valuable feature and  it's worth considering a more general-purpose auto complete implementation that can be used across different types of inputs.

### High contrast mode

> Be careful about visibility of the spin button arrows in high contrast mode.

Author: Todo added in the file

### Motion and animation

> Animations occur when the value changes to give the appearance of spinning. This needs to respond to user settings to turn off animations.

Reviewer 2: This isn't currently part of our v9 (or v8) SpinButton. Unless we have a plan to add this, I think we can leave this line off.

Placeholder: Mouse press speed of value changes visual indicators

### Semantics

## Known issues

Placeholder: If there are any current known bugs with browsers or AT that impact this component, list them here.

> Mouse press quick changes to value are not announced to screen readers.

Reviewer 2: This can go in a different PR, but I'm dropping this here for reference:

- NVDA doen't read the value of spinbuttons in Chromium: https://github.com/nvaccess/nvda/issues/13195
- Narrator defaults to reading min/max as 0 when they're (intentionally) undefined: https://microsoft.visualstudio.com/Edge/_workitems/edit/39070743

## Usage

### When to choose Spin button

SpinButtons allow someone to incrementally adjust a value in small steps.

### Required props

Placeholder: If there are any props or other specific work that the author must do to make the default usage of the component accessible, list those here. One example is that all form components must have a label provided by the author. If the component has any internally defined strings that should be localized, list those here.

Open question: How does everything read? Name, role, value? What about when there are prefixes or postfixes?

### Styling

### Advanced usage

Placeholder: Cover potential use cases not included in our storybook examples here (or even included storybook examples, if they're complex require a notable amout of work or nuanced understanding from authors)

#### Child content restrictions

The following child content is allowed:

> 1. Text

Reviewer 2: I believe we actually do not allow any children, since the primary slot is an `<input>` element.

We could mention that the up/down buttons can have children, though.

Author: Other than carats, what other children are allowed?

Reviewer: > Other than carats, what other children are allowed?

Technically, anything that can be a child of a `<button>`. In practice it should probably be restricted to a 16px icon from `@fluentui/react-icons` as other content will likely break the visual design.

## Known issues

Mouse press quick changes to value are not announced to screen readers.

[NVDA does not read the value of spinbuttons in Chromium](https://github.com/nvaccess/nvda/issues/13195)

[Narrator defaults to reading min/max as 0 when they're (intentionally) undefined](https://microsoft.visualstudio.com/Edge/_workitems/edit/39070743)

### Narrator + Edge

> When no min or max values are set Narrator announces "minimum 0" and "maximum 0". This is misleading because when these values are not set SpinButton does not enforce a min or max value. Example.

Reviewer: ```suggestion
When no min or max values are set Narrator announces "minimum 0" and "maximum 0". This is misleading because when these values are not set SpinButton does not enforce a min or max value.
```

### Required props

Authors must provide a label for SpinButton. The recommended pattern for Fluent inputs is to use the Label component like this:
```
<Label htmlFor="example-spinbutton">A SpinButton</Label>
<SpinButton id="example-spinbutton" defaultValue={10} />
```

How the SpinButton is read depends on the screen reader used but generally it reads:

1. Label
2. "spin button"
3. "value" where "value" is the current displayed value

> Prefixes, postfixes and other display formatted (e.g., mapping the numbers 1-12 to January-December) is handled by setting the `aria-valuetext` attribute.

Reviewer: ```suggestion
Prefixes, postfixes and other display formatting (e.g., mapping the numbers 1-12 to January-December) is handled by setting the `displayValue` prop which, internally to `SpinButton` sets the `aria-valuetext` attribute.
```
